Userland/Services/WebContent/BackingStoreManager.cpp

```diff
@@ -24,6 +45,71 @@ void BackingStoreManager::restart_resize_timer()
     m_backing_store_shrink_timer->restart();
 }
 
+void BackingStoreManager::reallocate_backing_stores(Gfx::IntSize size)
+{
+#ifdef AK_OS_MACOS
+    if (!s_iosurface_usage_disabled) {
+        auto back_iosurface = Core::IOSurfaceHandle::create(size.width(), size.height());
+        auto back_iosurface_port = back_iosurface.create_mach_port();
+
+        auto front_iosurface = Core::IOSurfaceHandle::create(size.width(), size.height());
+        auto front_iosurface_port = front_iosurface.create_mach_port();
+
+        m_front_bitmap_id = m_next_bitmap_id++;
+        m_back_bitmap_id = m_next_bitmap_id++;
+
+        m_front_bitmap = Gfx::Bitmap::create_wrapper(Gfx::BitmapFormat::BGRA8888, size, size.width() * front_iosurface.bytes_per_element(), front_iosurface.data(), [handle = move(front_iosurface)] {}).release_value();
+        m_back_bitmap = Gfx::Bitmap::create_wrapper(Gfx::BitmapFormat::BGRA8888, size, size.width() * back_iosurface.bytes_per_element(), back_iosurface.data(), [handle = move(back_iosurface)] {}).release_value();
+
+        Core::Platform::BackingStoreMetadata metadata;
+        metadata.page_id = m_page_client.m_id;
+        metadata.front_backing_store_id = m_front_bitmap_id;
+        metadata.back_backing_store_id = m_back_bitmap_id;
+
+        Core::Platform::MessageWithBackingStores message;
+
+        message.header.msgh_remote_port = s_browser_mach_port->port();
+        message.header.msgh_local_port = MACH_PORT_NULL;
+        message.header.msgh_bits = MACH_MSGH_BITS(MACH_MSG_TYPE_COPY_SEND, 0) | MACH_MSGH_BITS_COMPLEX;
```

Comment:
if i understand the rest of the code correctly, you create a send once right in IOSurfaceHandle, and then ask the kernel to COPY_SEND it? Is that correct? should we either: 1) create a send right instead of a send_once right or 2) COPY_SEND_ONCE (if that's a thing?)

Reply:
As said above, I'm pretty sure IOSurface's right is send, not send-once.

In this line here, we're setting the `MACH_MSG_TYPE_COPY_SEND` disposition for the _request port_ (`s_browser_mach_port->port()`); but down below we're also doing copy-send for the two IOSurface ports. So yes, they're clearly send rights, not send-onces.

> COPY_SEND_ONCE (if that's a thing?)

:rofl:  `MACH_MSG_TYPE_COPY_SEND_ONCE` was [an April Fools joke](https://mail.gnu.org/archive/html/bug-hurd/2024-04/msg00001.html) (that you [fell for](https://serenityos.social/@akaster/112196244361179044) :wink:). It's not a real thing, and it makes no sense to copy send-once rights.

Reply:
yes, using COPY_SEND_ONCE was a mistake, I replaced it with COPY_SEND. thanks for clarifying this!